File: DSA/DSA_450/05_linked_list/013_Merge_Sort_Linked_List.cpp
/*
Problem: Merge Sort for Linked List
URL: https://practice.geeksforgeeks.org/problems/sort-a-linked-list/1

Problem Statement:
Given Pointer/Reference to the head of the linked list, the task is to Sort the given linked list using Merge Sort.

Sample Input/Output:
Input: N = 5, value[] = {3,5,2,4,1}
Output: 1->2->3->4->5
Explanation: After sorting the given linked list, the resultant will be 1->2->3->4->5.
*/

#include <bits/stdc++.h>
using namespace std;

struct ListNode {
    int data;
    ListNode* next;
    ListNode(int x) : data(x), next(NULL) {}
};

ListNode* Create_List(vector<int>& arr) {
    if (arr.empty()) return NULL;
    ListNode* head = new ListNode(arr[0]);
    ListNode* curr = head;
    for (int i = 1; i < arr.size(); i++) {
        curr->next = new ListNode(arr[i]);
        curr = curr->next;
    }
    return head;
}

void Print_List(ListNode* head) {
    while (head) {
        cout << head->data;
        if (head->next) cout << "->";
        head = head->next;
    }
    cout << endl;
}

class Solution {
public:
    ListNode* Merge_Sort_Recursive(ListNode* head) {
        /*
        Merge Sort recursive with split
        Time Complexity: O(n log n)
        Space Complexity: O(log n)
        */
        if (!head || !head->next) return head;
        
        ListNode* mid = Get_Middle(head);
        ListNode* nextToMid = mid->next;
        mid->next = NULL;
        
        ListNode* left = Merge_Sort_Recursive(head);
        ListNode* right = Merge_Sort_Recursive(nextToMid);
        
        return Merge_Two_Sorted_Lists(left, right);
    }
    
    ListNode* Merge_Sort_Iterative(ListNode* head) {
        /*
        Iterative merge (bottom-up)
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        */
        if (!head || !head->next) return head;
        
        int length = Get_Length(head);
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        
        for (int size = 1; size < length; size *= 2) {
            ListNode* prev = dummy;
            ListNode* curr = dummy->next;
            
            while (curr) {
                ListNode* left = curr;
                ListNode* right = Split_List(left, size);
                curr = Split_List(right, size);
                
                prev->next = Merge_Two_Sorted_Lists(left, right);
                while (prev->next) {
                    prev = prev->next;
                }
            }
        }
        
        head = dummy->next;
        delete dummy;
        return head;
    }
    
private:
    ListNode* Get_Middle(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head->next;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
    
    ListNode* Merge_Two_Sorted_Lists(ListNode* list1, ListNode* list2) {
        ListNode* dummy = new ListNode(0);
        ListNode* tail = dummy;
        
        while (list1 && list2) {
            if (list1->data <= list2->data) {
                tail->next = list1;
                list1 = list1->next;
            } else {
                tail->next = list2;
                list2 = list2->next;
            }
            tail = tail->next;
        }
        
        tail->next = list1 ? list1 : list2;
        ListNode* result = dummy->next;
        delete dummy;
        return result;
    }
    
    int Get_Length(ListNode* head) {
        int len = 0;
        while (head) {
            len++;
            head = head->next;
        }
        return len;
    }
    
    ListNode* Split_List(ListNode* head, int n) {
        for (int i = 1; head && i < n; i++) {
            head = head->next;
        }
        if (!head) return NULL;
        ListNode* next = head->next;
        head->next = NULL;
        return next;
    }
};

void Test_Merge_Sort_Linked_List() {
    Solution solution;
    
    vector<int> arr = {3, 5, 2, 4, 1};
    ListNode* head = Create_List(arr);
    ListNode* result1 = solution.Merge_Sort_Recursive(head);
    cout << "Test 1 - Recursive: ";
    Print_List(result1);
    
    arr = {3, 5, 2, 4, 1};
    head = Create_List(arr);
    ListNode* result2 = solution.Merge_Sort_Iterative(head);
    cout << "Test 1 - Iterative: ";
    Print_List(result2);
    
    vector<int> arr2 = {4, 2, 1, 3};
    head = Create_List(arr2);
    result1 = solution.Merge_Sort_Recursive(head);
    cout << "Test 2 - Recursive: ";
    Print_List(result1);
}

int main() {
    Test_Merge_Sort_Linked_List();
    return 0;
}

File: DSA/DSA_450/Python/05_linked_list/013_Merge_Sort_Linked_List.py
"""
Problem: Merge Sort for Linked List
URL: https://practice.geeksforgeeks.org/problems/sort-a-linked-list/1

Problem Statement:
Given Pointer/Reference to the head of the linked list, the task is to Sort the given linked list using Merge Sort.

Sample Input/Output:
Input: N = 5, value[] = {3,5,2,4,1}
Output: 1->2->3->4->5
Explanation: After sorting the given linked list, the resultant will be 1->2->3->4->5.
"""

class ListNode:
    def __init__(self, x):
        self.data = x
        self.next = None

def Create_List(arr):
    if not arr:
        return None
    head = ListNode(arr[0])
    curr = head
    for i in range(1, len(arr)):
        curr.next = ListNode(arr[i])
        curr = curr.next
    return head

def List_To_Array(head):
    result = []
    while head:
        result.append(head.data)
        head = head.next
    return result

def Print_List(head):
    arr = List_To_Array(head)
    print("->".join(map(str, arr)))

class Solution:
    def Get_Middle(self, head):
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow
    
    def Merge_Two_Sorted_Lists(self, list1, list2):
        dummy = ListNode(0)
        tail = dummy
        
        while list1 and list2:
            if list1.data <= list2.data:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
        
        tail.next = list1 if list1 else list2
        return dummy.next
    
    def Merge_Sort_Recursive(self, head):
        """
        Merge Sort recursive with split
        Time Complexity: O(n log n)
        Space Complexity: O(log n)
        """
        if not head or not head.next:
            return head
        
        mid = self.Get_Middle(head)
        next_to_mid = mid.next
        mid.next = None
        
        left = self.Merge_Sort_Recursive(head)
        right = self.Merge_Sort_Recursive(next_to_mid)
        
        return self.Merge_Two_Sorted_Lists(left, right)
    
    def Get_Length(self, head):
        length = 0
        while head:
            length += 1
            head = head.next
        return length
    
    def Split_List(self, head, n):
        for i in range(1, n):
            if head:
                head = head.next
        if not head:
            return None
        next_node = head.next
        head.next = None
        return next_node
    
    def Merge_Sort_Iterative(self, head):
        """
        Iterative merge (bottom-up)
        Time Complexity: O(n log n)
        Space Complexity: O(1)
        """
        if not head or not head.next:
            return head
        
        length = self.Get_Length(head)
        dummy = ListNode(0)
        dummy.next = head
        
        size = 1
        while size < length:
            prev = dummy
            curr = dummy.next
            
            while curr:
                left = curr
                right = self.Split_List(left, size)
                curr = self.Split_List(right, size) if right else None
                
                prev.next = self.Merge_Two_Sorted_Lists(left, right)
                while prev.next:
                    prev = prev.next
            
            size *= 2
        
        return dummy.next

def Test_Merge_Sort_Linked_List():
    solution = Solution()
    
    arr = [3, 5, 2, 4, 1]
    head = Create_List(arr)
    result1 = solution.Merge_Sort_Recursive(head)
    print("Test 1 - Recursive: ", end="")
    Print_List(result1)
    
    arr = [3, 5, 2, 4, 1]
    head = Create_List(arr)
    result2 = solution.Merge_Sort_Iterative(head)
    print("Test 1 - Iterative: ", end="")
    Print_List(result2)
    
    arr2 = [4, 2, 1, 3]
    head = Create_List(arr2)
    result1 = solution.Merge_Sort_Recursive(head)
    print("Test 2 - Recursive: ", end="")
    Print_List(result1)

if __name__ == "__main__":
    Test_Merge_Sort_Linked_List()
